woocommerce/google-listings-and-ads
# Confirm possible overwrite of existing product feeds

### Changes proposed in this Pull Request:

Backend part of #280.

This PR adds a check for existing product feeds during the Merchant Center onboarding process, specifically in the `set_id` step, and after checking/updating the website URL. If there are existing product feeds, a `409` is returned and the user should consent to possible feed changes/overwriting by making a `POST` request to the `/mc/accounts/feed-overwrite` endpoint with the body `{"id":...}`. 

This endpoint works the same as `switch-url` and `claim-overwrite`, essentially bypassing the error and continuing with the onboarding process. Any MC account set up correctly will have user consent to overwrite or change existing product feeds.

**Notes**
- This PR won't be merged until the front end is prepared for the slight modifications it makes to the MC account setup flow.
- Content API feeds aren't displayed in the [`datafeeds.list`](https://developers.google.com/shopping-content/reference/rest/v2.1/datafeeds/list) results, which means that the Merchant Center _COULD_ have another source uploading/syncing products using the Content API, and it wouldn't be flagged here. A possible workarounds would be retrieving the existing products and looking at the `source` property (it would be `api`).
- No actions are actually taken on the existing feeds during the onboarding process. If they should be deleted or disabled, that would require further changes (either to this step of the onboarding process, or before product sync). This change simply ensures that the user consents to possible future changes in order to proceed with the Merchant Center onboarding.
- This adds a new attribute to the `409` response that's shared by the URL Conflict and the Existing Feeds errors: `action`. This attribute indicates which endpoint should be called next in order to confirm user consent for overwrite/switch. 
- It's possible for a user to have *BOTH* a conflicting website URL *AND* existing product feeds. In this case, a '409' error would be returned twice:
    - After the initial `POST`, with `"action":"switch-url"`
    - And after the subsequent `POST` to `/switch-url`, with `"action":"feed-overwrite"`. 
    It would be necessary to call both overwrite endpoints in order to continue the setup process. 
    On the front end, the user would have to click on two confirmation buttons: first to accept the URL switch and then to consent to feed changes/overwriting.

**Follow-ups**
- [ ] Confirm whether we should actually disable or delete existing product feeds as part of the onboarding process.
- [ ] Confirm whether it's necessary to check for existing products uploaded via the Content API.
- [ ] Update the front end to check the `action` attribute when a `409` is returned.
- [ ] Determine whether the existing product feeds check should be skipped if we already updated the MC account's URL (presumably their feeds would fail because the existing feeds should be uploading products that point to the old URL).
- [ ] Refactor AccountController (?)

### Detailed test instructions:

1. In an MC account with a claimed website URL, create a new primary product feed (_not_ "Content API" input method): <https://merchants.google.com/mc/products/sources/createPrimaryFeed>
2. Run the MC Account Setup process from Connection Test using the existing MC account ID.
3. Confirm a `409` response with the message indicating existing feeds, and `"action":"feed-overwrite"`.
4. Click the "Feed Overwrite" button on Connection Test and confirm that the MC account setup process completes correctly.
Bonus: Same, but with an MC account with a different claimed website URL (will require clicking "Switch URL" button first, and then "Feed Overwrite" button


### Changelog Note:

> Check for existing product feeds and require user consent to change/overwrite them in order to continue the Merchant Center account setup process.
